azkaban-common/src/main/java/azkaban/executor/dispatcher/CandidateComparator.java

```diff
@@ -0,0 +1,187 @@
+
+  // internal repository of the registered comparators .
+  private Map<String,FactorComparator<T>> factorComparatorList =
+      new ConcurrentHashMap<String,FactorComparator<T>>();
+
+  /** gets the name of the current implementation of the candidate comparator.
+   * @returns : name of the comparator.
+   * */
+  public abstract String getName();
+
+  /** differentiate method which will kick in when the comparator list generated an equality result for
+   *  both sides. the differentiate method will try best to make sure a stable result is returned.
+   * */
+  protected boolean differentiate(T object1, T object2){
+    if (object2 == null) return true;
+    if (object1 == null) return false;
+
+    return object1.hashCode() >= object2.hashCode();
+  }
+
+  /** function to register a factorComparator to the internal Map for future reference.
+   * @param factorComparator : the comparator object to be registered.
+   * */
+  protected void registerFactorComparator(FactorComparator<T> comparator){
+      if (null == comparator ||
+          Integer.MAX_VALUE - this.getTotalWeight() < comparator.getWeight() ) {
+        logger.info(
+            "skipping registerFactorComparator as the comaractor is null or has an invalid weight value.");
+        return;
+      }
+
+      // add or replace the Comparator.
+      this.factorComparatorList.put(comparator.getFactorName(),comparator);
+      logger.info(String.format("Factor comparator added for '%s'. Weight = '%s'",
+          comparator.getFactorName(), comparator.getWeight()));
+  }
+
+  /** function update the weight of a specific registered factorCompartor.
+   * @param factorName : the name of the registered factorComparator to adjust.
+   * @param weight:      the new weight value to be adjusted to.
+   * @return -1 if the factor doesn't exist or the weight value specified is invalid,
+   *          the original value before update otherwise.
+   * */
+  public int adjustFactorWeight(String factorName, int weight){
+    // shortcut if the input is invalid.
+    if (factorName == null ||
+        factorName == "" ||
+        weight < 0 ||
+        Integer.MAX_VALUE - this.getTotalWeight() < weight){
+      logger.info(
+          "skipping adjustFactorWeight as one or more of the input parameters are invalid");
+      return -1;
+    }
+
+    FactorComparator<T> value = this.factorComparatorList.get(factorName);
+
+    // shortcut if the key doesn't exist.
+    if (null == value){
+      logger.info(String.format(
+          "unable to udpate weight as the specified factorName %s doesn't exist",
+          factorName));
+      return -1;
+    }
+
+    int returnVal = value.getWeight();
+    value.updateWeight(weight);
+    return returnVal;
+  }
+
+  /** function returns the total weight of the registered comparators.
+   * @return the value of total weight.
+   * */
+  public int getTotalWeight(){
+    int totalWeight = 0 ;
```

Comment:
hum.... the code actually doesn't cache it and does compute every time the method is called. And the reason not to cache is that we have the function to adjust the weight so it will be a little bit messy when the weight is changed and we need to update the cache.  please let me know if I get it right .